README.md
ZXCVBN-Perl
==========

A realistic password strength estimator.

This is a Perl implementation of the library created by the team at Dropbox,
also inspired in [zxcvbn-python](https://github.com/dwolfhub/zxcvbn-python).
The original library, written for JavaScript, can be found
[here](https://github.com/dropbox/zxcvbn).


INSTALLATION
-----------

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION
----------------------

After installing, you can find documentation for this module with the
perldoc command.

    perldoc ZXCVBN

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=ZXCVBN-Perl

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/ZXCVBN-Perl

    CPAN Ratings
        http://cpanratings.perl.org/d/ZXCVBN-Perl

    Search CPAN
        http://search.cpan.org/dist/ZXCVBN-Perl/


LICENSE AND COPYRIGHT
-------------------

Copyright (C) 2017 José María Fernández, BSC-CNS

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this program; if not, write to the Free
Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA



Changes
Revision history for ZXCVBN-Perl

0.01    Date/time
        First version, released on an unsuspecting world.



lib/ZXCVBN/Scoring.pm
#!/usr/bin/perl

use 5.014;
use feature 'unicode_strings';
use strict;
use warnings;

package ZXCVBN::Scoring;

use ZXCVBN::AdjacencyGraphs;
use boolean qw();
#use List::Util qw();

sub calc_average_degree(\%) {
	my($p_graph) = @_;
	
	my $average = 0;
	
	while(my($key,$p_neighbors) = each(%{$p_graph})) {
		$average += scalar(grep { defined($_) } @{$p_neighbors});
	}
	$average /= scalar(keys(%{$p_graph}));
	
	return $average;
}

use constant {
	BRUTEFORCE_CARDINALITY => 10,
	MIN_GUESSES_BEFORE_GROWING_SEQUENCE => 10000,
	MIN_SUBMATCH_GUESSES_SINGLE_CHAR => 10,
	MIN_SUBMATCH_GUESSES_MULTI_CHAR => 50,

	MIN_YEAR_SPACE => 20,
	REFERENCE_YEAR => 2017
};


=head2 nCk

http://blog.plover.com/math/choose.html
=cut
sub nCk($$) {
	my($n,$k) = @_;
	
	if($k > $n) {
		return 0;
	}
	if($k == $n) {
		return 1;
	}
	
	my $r = 1;
	foreach my $d (1..$k) {
		$r *= $n;
		$r /= $d;
		$n--;
	}
	
	return $r;
}


sub _factorial($) {
	my($l) = @_;
	return $l  if($l<=2);
	
	my $retval = 1;
	foreach my $v (2..$l) {
		$retval *= $v;
	}
	
	return $retval;
}

sub _max($$) {
	return $_[0] > $_[1] ? $_[0] : $_[1];
}

sub _min($$) {
	return $_[0] < $_[1] ? $_[0] : $_[1];
}


my $START_UPPER = qr/^[A-Z][^A-Z]+$/;
my $END_UPPER = qr/^[^A-Z]+[A-Z]$/;
my $ALL_UPPER = qr/^[^a-z]+$/;
my $ALL_LOWER = qr/^[^A-Z]+$/;


sub uppercase_variations(\%) {
	my($p_match) = @_;
	my $word = $p_match->{'token'};
	
	if($word =~ $ALL_LOWER || lc($word) eq $word) {
		return 1;
	}
	
	foreach my $regex ($START_UPPER, $END_UPPER, $ALL_UPPER) {
		if($word =~ $regex) {
			return 2;
		}
	}
	
	my $U = scalar(grep { $_ =~ /[[:upper:]]/ } split(//,$word));
	my $L = scalar(grep { $_ =~ /[[:lower:]]/ } split(//,$word));
	my $variations = 0;
	foreach my $i (1 .. _min($U, $L)) {
		$variations += nCk($U + $L, $i);
	}
	
	return $variations;
}


sub l33t_variations(\%) {
	my($p_match) = @_;
	
	return 1  unless(exists($p_match->{'l33t'}));
	
	my $variations = 1;
	
	while(my($subbed,$unsubbed) = each(%{$p_match->{'sub'}})) {
		# lower-case match.token before calculating: capitalization shouldn't
		# affect l33t calc.
		my @chrs = split(//,lc($p_match->{'token'}));
		my $S = scalar(grep { $_ eq $subbed } @chrs);
		my $U = scalar(grep { $_ eq $unsubbed } @chrs);
		if($S==0 || $U==0) {
			# for this sub, password is either fully subbed (444) or fully
			# unsubbed (aaa) treat that as doubling the space (attacker needs
			# to try fully subbed chars in addition to unsubbed.)
			$variations *= 2;
		} else {
			# this case is similar to capitalization:
			# with aa44a, U = 3, S = 2, attacker needs to try unsubbed + one
			# sub + two subs
			my $p = _min($U,$S);
			my $possibilities = 0;
			foreach my $i (1..$p) {
				$possibilities += nCk($U + $S, $i);
			}
			$variations *= $possibilities;
		}
	}
	
	return $variations;
}



sub bruteforce_guesses(\%) {
	my($p_match) = @_;
	my $guesses = BRUTEFORCE_CARDINALITY ** length($p_match->{'token'});
	# small detail: make bruteforce matches at minimum one guess bigger than
	# smallest allowed submatch guesses, such that non-bruteforce submatches
	# over the same [i..j] take precedence.
	my $min_guesses = undef;
	if(length($p_match->{'token'}) == 1) {
		$min_guesses = MIN_SUBMATCH_GUESSES_SINGLE_CHAR + 1;
	} else {
		$min_guesses = MIN_SUBMATCH_GUESSES_MULTI_CHAR + 1;
	}
	
	return _max($guesses, $min_guesses);
}

sub dictionary_guesses(\%) {
	my($p_match) = @_;
	
	# keep these as properties for display purposes
	$p_match->{'base_guesses'} = $p_match->{'rank'};
	$p_match->{'uppercase_variations'} = uppercase_variations(%{$p_match});
	$p_match->{'l33t_variations'} = l33t_variations(%{$p_match});
	my $reversed_variations = exists($p_match->{'reversed'}) ? 2 : 1;
	
	return $p_match->{'base_guesses'} * $p_match->{'uppercase_variations'} * \
		$p_match->{'l33t_variations'} * $reversed_variations;
}


sub repeat_guesses(\%) {
	my($p_match) = @_;
	
	return $p_match->{'base_guesses'} * $p_match->{'repeat_count'};
}


sub sequence_guesses(\%) {
	my($p_match) = @_;
	
	my $first_chr = substr($p_match->{'token'},0,1);
	# lower guesses for obvious starting points
	my $base_guesses = undef;
	if($first_chr ~~ ['a', 'A', 'z', 'Z', '0', '1', '9']) {
		$base_guesses = 4;
	} elsif($first_chr =~ /[[:digit:]]/) {
		$base_guesses = 10;  # digits
        } else {
		# could give a higher base for uppercase,
		# assigning 26 to both upper and lower sequences is more
		# conservative.
		$base_guesses = 26;
	}
	unless($p_match->{'ascending'}) {
		$base_guesses *= 2;
	}
        
	return $base_guesses * length($p_match->{'token'});
}

my %char_class_bases = (
        'alpha_lower'	=>	26,
        'alpha_upper'	=>	26,
        'alpha'	=>	52,
        'alphanumeric'	=>	62,
        'digits'	=>	10,
        'symbols'	=>	33,
);

sub regex_guesses(\%) {
	my($p_match) = @_;
	
	if(exists($char_class_bases{$p_match->{'regex_name'}})) {
		return $char_class_bases{$p_match->{'regex_name'}} ** length($p_match->{'token'});
	}
	my $year_space = undef;
	if($p_match->{'regex_name'} eq 'recent_year') {
		# conservative estimate of year space: num years from REFERENCE_YEAR.
		# if year is close to REFERENCE_YEAR, estimate a year space of
		# MIN_YEAR_SPACE.
		$year_space = int(abs(int($p_match->{'token'}) - REFERENCE_YEAR));
		$year_space = _max($year_space, MIN_YEAR_SPACE);
	}

        return $year_space;
}


sub date_guesses(\%) {
	my($p_match) = @_;
	
	my $year_space = _max(abs($p_match->{'year'} - REFERENCE_YEAR), MIN_YEAR_SPACE);
	my $guesses = $year_space * 365;
	if(exists($p_match->{'separator'})) {
		$guesses *= 4;
	}
        
	return $guesses;
}


my $KEYBOARD_AVERAGE_DEGREE = calc_average_degree(%{$ZXCVBN::AdjacencyGraphs::ADJACENCY_GRAPHS{'qwerty'}});
# slightly different for keypad/mac keypad, but close enough
my $KEYPAD_AVERAGE_DEGREE = calc_average_degree(%{$ZXCVBN::AdjacencyGraphs::ADJACENCY_GRAPHS{'keypad'}});

my $KEYBOARD_STARTING_POSITIONS = scalar(keys(%{$ZXCVBN::AdjacencyGraphs::ADJACENCY_GRAPHS{'qwerty'}}));
my $KEYPAD_STARTING_POSITIONS = scalar(keys(%{$ZXCVBN::AdjacencyGraphs::ADJACENCY_GRAPHS{'keypad'}}));


sub spatial_guesses(\%) {
	my($p_match) = @_;

	my $s = undef;
	my $d = undef;
	if($p_match->{'graph'} ~~ ['qwerty', 'dvorak']) {
		$s = $KEYBOARD_STARTING_POSITIONS;
		$d = $KEYBOARD_AVERAGE_DEGREE;
	} else {
		$s = $KEYPAD_STARTING_POSITIONS;
		$d = $KEYPAD_AVERAGE_DEGREE;
	}
	my $guesses = 0;
	my $L = length($p_match->{'token'});
	my $t = $p_match->{'turns'};
	# estimate the number of possible patterns w/ length L or less with t turns
	# or less.
	foreach my $i (2 .. $L) {
		my $possible_turns = _min($t, $i - 1) + 1;
		foreach my $j (1 .. ($possible_turns - 1)) {
			$guesses += nCk($i - 1, $j - 1) * $s * ($d ** $j);
		}
	}
	# add extra guesses for shifted keys. (% instead of 5, A instead of a.)
	# math is similar to extra guesses of l33t substitutions in dictionary
	# matches.
	if($p_match->{'shifted_count'}) {
		my $S = $p_match->{'shifted_count'};
		my $U = length($p_match->{'token'}) - $S;  # unshifted count
		if($S == 0 || $U == 0) {
		    $guesses *= 2;
		} else {
			my $shifted_variations = 0;
			foreach my $i (1, _min($S, $U)) {
				$shifted_variations += nCk($S + $U, $i);
			}
			$guesses *= $shifted_variations;
		}
	}

	return $guesses;
}


my %estimation_functions = (
	'bruteforce'	=> \&bruteforce_guesses,
	'dictionary'	=> \&dictionary_guesses,
	'spatial'	=> \&spatial_guesses,
	'repeat'	=> \&repeat_guesses,
	'sequence'	=> \&sequence_guesses,
	'regex'		=> \&regex_guesses,
	'date'		=> \&date_guesses
);


sub estimate_guesses(\%$) {
	my($p_match,$password) = @_;
	
	if(exists($p_match->{'guesses'})) {
		return $p_match->{'guesses'};
	}
	
	my $min_guesses = 1;
	if(length($p_match->{'token'}) < length($password)) {
		if(length($p_match->{'token'}) == 1) {
			$min_guesses = MIN_SUBMATCH_GUESSES_SINGLE_CHAR;
		} else {
			$min_guesses = MIN_SUBMATCH_GUESSES_MULTI_CHAR;
		}
	}
	
	my $guesses = $estimation_functions{$p_match->{'pattern'}}->($p_match);
	$p_match->{'guesses'} = _max($guesses, $min_guesses);
	$p_match->{'guesses_log10'} = log($p_match->{'guesses'}) / log(10);

	return $p_match->{'guesses'};
}

# helper: considers whether a length-l sequence ending at match m is better
# (fewer guesses) than previously encountered sequences, updating state if
# so.
sub _update(\%$$\%$) {
	my($p_m, $l, $password,$p_optimal,$_exclude_additive) = @_;
	
	my $k = $p_m->{'j'};
	my $pi = estimate_guesses(%{$p_m}, $password);
	if($l > 1) {
		# we're considering a length-l sequence ending with match m:
		# obtain the product term in the minimization function by
		# multiplying m's guesses by the product of the length-(l-1)
		# sequence ending just before m, at m.i - 1.
		$pi *= $p_optimal->{'pi'}[$p_m->{'i'} - 1]{$l - 1};
	}
	# calculate the minimization func
	my $g = _factorial($l) * $pi;
	unless($_exclude_additive) {
		$g += MIN_GUESSES_BEFORE_GROWING_SEQUENCE ** ($l - 1);
	}

	# update state if new best.
	# first see if any competing sequences covering this prefix, with l or
	# fewer matches, fare better than this sequence. if so, skip it and
	# return.
	while(my($competing_l,$competing_g) = each(%{$p_optimal->{'g'}[$k]})) {
		next  if($competing_l > $l);
		return  if($competing_g <= $g);
	}
	
	# this sequence might be part of the final optimal sequence.
	$p_optimal->{'g'}[$k]{$l} = $g;
	$p_optimal->{'m'}[$k]{$l} = $p_m;
	$p_optimal->{'pi'}[$k]{$l} = $pi;
}

# helper: make bruteforce match objects spanning i to j, inclusive.
sub _make_bruteforce_match($$$) {
	my($i, $j, $password) = @_;
	return {
		'pattern'	=>	'bruteforce',
		'token'	=>	substr($password,$i,$j-$i+1),
		'i'	=>	$i,
		'j'	=>	$j,
	};
}

# helper: evaluate bruteforce matches ending at k.
sub _bruteforce_update($$\%$) {
	my($k,$password,$p_optimal, $_exclude_additive) = @_;
	
        # see if a single bruteforce match spanning the k-prefix is optimal.
        my $p_m = _make_bruteforce_match(0, $k, $password);
        _update(%{$p_m}, 1, $password, %{$p_optimal}, $_exclude_additive);
        foreach my $i (1 .. $k) {
		# generate k bruteforce matches, spanning from (i=1, j=k) up to
		# (i=k, j=k). see if adding these new matches to any of the
		# sequences in optimal[i-1] leads to new bests.
		my $p_m = _make_bruteforce_match($i, $k, $password);
		while(my($l, $last_m) = each(%{$p_optimal->{'m'}[$i - 1]})) {
			$l = int($l);

			# corner: an optimal sequence will never have two adjacent
			# bruteforce matches. it is strictly better to have a single
			# bruteforce match spanning the same region: same contribution
			# to the guess product with a lower length.
			# --> safe to skip those cases.
			if(exists($last_m->{'pattern'}) && $last_m->{'pattern'} eq 'bruteforce') {
				next;
			}
			
			# try adding m to this length-l sequence.
			_update(%{$p_m}, $l + 1, $password, %{$p_optimal}, $_exclude_additive);
		}
	}
}

# helper: step backwards through optimal.m starting at the end,
# constructing the final optimal match sequence.
sub _unwind($\%) {
	my($n,$p_optimal) = @_;
	
	my @optimal_match_sequence = ();
	
	my $k = $n - 1;
	# find the final best sequence length and score
	my $l = undef;
	my $g = 'inf' + 0;
	while(my($candidate_l,$candidate_g) = each(%{$p_optimal->{'g'}[$k]})) {
		if($candidate_g < $g) {
			$l = $candidate_l;
			$g = $candidate_g;
		}
	}
	
	while($k >= 0) {
		my $p_m = $p_optimal->{'m'}[$k]{$l};
		unshift(@optimal_match_sequence, $p_m);
		$k = $p_m->{'i'} - 1;
		$l -= 1;
	}

	return \@optimal_match_sequence;
}

=head2 most_guessable_match_sequence

------------------------------------------------------------------------------
search --- most guessable match sequence -------------------------------------
------------------------------------------------------------------------------

takes a sequence of overlapping matches, returns the non-overlapping sequence with
minimum guesses. the following is a O(l_max * (n + m)) dynamic programming algorithm
for a length-n password with m candidate matches. l_max is the maximum optimal
sequence length spanning each prefix of the password. In practice it rarely exceeds 5 and the
search terminates rapidly.

the optimal "minimum guesses" sequence is here defined to be the sequence that
minimizes the following function:

   g = l! * Product(m.guesses for m in sequence) + D^(l - 1)

where l is the length of the sequence.

the factorial term is the number of ways to order l patterns.

the D^(l-1) term is another length penalty, roughly capturing the idea that an
attacker will try lower-length sequences first before trying length-l sequences.

for example, consider a sequence that is date-repeat-dictionary.
 - an attacker would need to try other date-repeat-dictionary combinations,
   hence the product term.
 - an attacker would need to try repeat-date-dictionary, dictionary-repeat-date,
   ..., hence the factorial term.
 - an attacker would also likely try length-1 (dictionary) and length-2 (dictionary-date)
   sequences before length-3. assuming at minimum D guesses per pattern type,
   D^(l-1) approximates Sum(D^i for i in [1..l-1]

------------------------------------------------------------------------------
=cut
sub most_guessable_match_sequence($\@;$) {
	my($password,$p_matches,$_exclude_additive) = @_;
	
	my $n = length($password);
	
	# partition matches into sublists according to ending index j
	my @matches_by_j = map { [] } (1..$n);
	foreach my $p_m (@{$p_matches}) {
		push(@{$matches_by_j[$p_m->['j']]},$p_m)  if(exists($p_m->['j']));
	}
	# small detail: for deterministic output, sort each sublist by i.
	foreach my $p_lst (@matches_by_j) {
		@{$p_lst} = sort { $a->{'i'} <=> $b->{'j'} } @{$p_lst};
	}
	
	my %optimal = (
		# optimal.m[k][l] holds final match in the best length-l match sequence
		# covering the password prefix up to k, inclusive.
		# if there is no length-l sequence that scores better (fewer guesses)
		# than a shorter match sequence spanning the same prefix,
		# optimal.m[k][l] is undefined.
		'm'	=> [ map { {} } (1..$n) ],

		# same structure as optimal.m -- holds the product term Prod(m.guesses
		# for m in sequence). optimal.pi allows for fast (non-looping) updates
		# to the minimization function.
		'pi'	=> [ map { {} } (1..$n) ],

		# same structure as optimal.m -- holds the overall metric.
		'g'	=> [ map { {} } (1..$n) ]
	);
	
	foreach my $k (0..$n-1) {
		foreach my $p_m (@{$matches_by_j[$k]}) {
			if($p_m->{'i'} >0) {
				foreach my $l (@{$optimal{'m'}[$p_m->{'i'} - 1]}) {
					$l = int($l);
					_update(%{$p_m}, $l + 1, $password, %optimal, $_exclude_additive);
				}
			} else {
				_update(%{$p_m}, 1, $password, %optimal, $_exclude_additive);
			}
		}
		_bruteforce_update($k,$password,%optimal, $_exclude_additive);
	}
	
	my $p_optimal_match_sequence = _unwind($n,%optimal);
	my $optimal_l = scalar(@{$p_optimal_match_sequence});
	
	# corner: empty password
	my $guesses = undef;
	if(length($password) == 0) {
		$guesses = 1;
	} else {
		$guesses = $optimal{'g'}[$n - 1]{$optimal_l};
	}
	
	# final result object
	return {
		'password'	=>	$password,
		'guesses'	=>	$guesses,
		'guesses_log10'	=>	log($guesses) / log(10),
		'sequence'	=>	$p_optimal_match_sequence
	};
}

1;


lib/ZXCVBN/Feedback.pm
#!/usr/bin/perl

use 5.006;
use feature 'unicode_strings';
use strict;
use warnings;

package ZXCVBN::Feedback;

use ZXCVBN::Scoring;

sub get_dictionary_match_feedback(\%$) {
	my($p_match,$is_sole_match) = @_;
	
	my $warning = '';
	if($p_match->{'dictionary_name'} eq 'passwords')  {
		if($is_sole_match && !exists($p_match->{'l33t'}) && !$p_match->{'reversed'}) {
			if($p_match->{'rank'} <= 10) {
				$warning = 'This is a top-10 common password.';
			} elsif($p_match->{'rank'} <= 100) {
				$warning = 'This is a top-100 common password.';
			} else {
				$warning = 'This is a very common password.';
			}
		} elsif($p_match->{'guesses_log10'} <= 4) {
			$warning = 'This is similar to a commonly used password.';
		}
	} elsif($p_match->{'dictionary_name'} eq 'english') {
		if($is_sole_match) {
			$warning = 'A word by itself is easy to guess.';
		}
	} elsif($p_match->{'dictionary_name'} ~~ ['surnames', 'male_names', 'female_names']) {
		if($is_sole_match) {
			$warning = 'Names and surnames by themselves are easy to guess.';
		} else {
			$warning = 'Common names and surnames are easy to guess.';
		}
	}
	
	my @suggestions = ();
	my $word = $p_match->{'token'};
	if($word =~ $ZXCVBN::Scoring::START_UPPER) {
		push(@suggestions,"Capitalization doesn't help very much.");
	} elsif($word =~ $ZXCVBN::Scoring::ALL_UPPER && lc($word) ne $word) {
		push(@suggestions,"All-uppercase is almost as easy to guess as all-lowercase.");
	}
	
	if($p_match->{'reversed'} && length($word) >= 4) {
		push(@suggestions,"Reversed words aren't much harder to guess.");
	}
	if(exists($p_match->{'l33t'})) {
		push(@suggestions,"Predictable substitutions like '\@' instead of 'a' don't help very much.");
	}
	
	return {
		'warning'	=> $warning,
		'suggestions'	=> \@suggestions
	};
}

sub get_match_feedback(\%$) {
	my($p_match,$is_sole_match) = @_;
	
	if($p_match->{'pattern'} eq 'dictionary') {
		return get_dictionary_match_feedback(%{$p_match},$is_sole_match);
	} elsif($p_match->{'pattern'} eq 'spatial') {
		my $warning = undef;
		if($p_match->{'turns'} == 1) {
			$warning = 'Straight rows of keys are easy to guess.';
		} else {
			$warning = 'Short keyboard patterns are easy to guess.';
		}
		
		return {
			'warning'	=> $warning,
			'suggestions'	=> [
				'Use a longer keyboard pattern with more turns.'
			]
		};
	} elsif($p_match->{'pattern'} eq 'repeat') {
		my $warning = undef;
		if(length($p_match->{'base_token'}) == 1) {
			$warning = 'Repeats like "aaa" are easy to guess.';
		} else {
			$warning = 'Repeats like "abcabcabc" are only slightly harder to guess than "abc".';
		}
		
		return {
			'warning'	=> $warning,
			'suggestions'	=> [
				'Avoid repeated words and characters.'
			]
		};
	} elsif($p_match->{'pattern'} eq 'sequence') {
		return {
			'warning'	=> 'Sequences like "abc" or "6543" are easy to guess.',
			'suggestions'	=> [
				'Avoid sequences.'
			]
		};
	} elsif($p_match->{'pattern'} eq 'regex') {
		if($p_match->{'regex_name'} eq 'recent_year') {
			return {
				'warning'	=> 'Recent years are easy to guess.',
				'suggestions'	=> [
					'Avoid recent years.',
					'Avoid years that are associated with you.'
				]
			};
		}
	} elsif($p_match->{'pattern'} eq 'date') {
		return {
			'warning'	=> 'Dates are often easy to guess.',
			'suggestions'	=> [
				'Avoid dates and years that are associated with you.',
			]
		}
	}
	
	return undef;
}

sub get_feedback($\@) {
	my($score,$p_sequence) = @_;
	
	if(scalar(@{$p_sequence}) == 0) {
		return {
			'warning'	=> '',
			'suggestions'	=> [
				"Use a few words, avoid common phrases.",
				"No need for symbols, digits, or uppercase letters."
			]
		};
	}
	
	if($score > 2) {
		return {
			'warning'	=> '',
			'suggestions'	=> [ ]
		};
	}
	
	my $longest_match = undef;
	foreach my $p_match (@{$p_sequence}) {
		if(!defined($longest_match) || length($p_match->{'token'}) > length($longest_match->{'token'})) {
			$longest_match = $p_match;
		}
	}
	
	my $p_feedback = get_match_feedback(%{$longest_match},scalar(@{$p_sequence}) == 1);
	my $extra_feedback = 'Add another word or two. Uncommon words are better.';
	if(defined($p_feedback)) {
		unshift(@{$p_feedback->{'suggestions'}},$extra_feedback);
		$p_feedback->{'warning'} = ''  unless(exists($p_feedback->{'warning'}));
	} else {
		$p_feedback = {
			'warning'	=> '',
			'suggestions'	=> [ $extra_feedback ]
		};
	}
	
	return $p_feedback;
}


1;


lib/ZXCVBN/TimeEstimates.pm
#!/usr/bin/perl

use 5.006;
use feature 'unicode_strings';
use strict;
use warnings;

package ZXCVBN::TimeEstimates;

use POSIX qw();

use constant DELTA	=> 5;

sub guesses_to_score($) {
	my($guesses) = @_;
	
	if($guesses < 1e3 + DELTA) {
		# risky password: "too guessable"
		return 0;
	} elsif($guesses < 1e6 + DELTA) {
		# modest protection from throttled online attacks: "very guessable"
		return 1;
	} elsif($guesses < 1e8 + DELTA) {
		# modest protection from unthrottled online attacks: "somewhat
		# guessable"
		return 2;
	} elsif($guesses < 1e10 + DELTA) {
		# modest protection from offline attacks: "safely unguessable"
		# assuming a salted, slow hash function like bcrypt, scrypt, PBKDF2,
		# argon, etc
		return 3;
	} else {
		# strong protection from offline attacks under same scenario: "very
		# unguessable"
		return 4;
	}
}

use constant MINUTE	=> 60.0;
use constant HOUR	=> MINUTE * 60.0;
use constant DAY	=> HOUR * 24.0;
use constant MONTH	=> DAY * 31.0;
use constant YEAR	=> MONTH * 12.0;
use constant CENTURY	=> YEAR * 100.0;

sub display_time($) {
	my($seconds) = @_;
	
	my $display_num = undef;
	my $display_str = undef;
	if($seconds < 1) {
		#$display_num = undef;
		$display_str = 'less than a second';
	} elsif($seconds < MINUTE) {
		my $base = POSIX::round($seconds);
		$display_num = $base;
		$display_str = $base .' second';
	} elsif($seconds < HOUR) {
		my $base = POSIX::round($seconds / MINUTE);
		$display_num = $base;
		$display_str = $base .' minute';
	} elsif($seconds < DAY) {
		my $base = POSIX::round($seconds / HOUR);
		$display_num = $base;
		$display_str = $base .' hour';
	} elsif($seconds < MONTH) {
		my $base = POSIX::round($seconds / DAY);
		$display_num = $base;
		$display_str = $base . ' day';
	} elsif($seconds < YEAR) {
		my $base = POSIX::round($seconds / MONTH);
		$display_num = $base;
		$display_str = $base . ' month';
	} elsif($seconds < CENTURY) {
		my $base = POSIX::round($seconds / YEAR);
		$display_num = $base;
		$display_str = $base . ' year';
	} else {
		#$display_num = undef;
		$display_str = 'centuries';
	}
	
	if(defined($display_num) && $display_num != 1) {
		$display_str .= 's';
	}
	
	return $display_str;
}

sub estimate_attack_times($) {
	my($guesses) = @_;
	
	# Translating it into a floating point number
	$guesses = $guesses + 0.0;
	
	my %crack_times_seconds = (
		'online_throttling_100_per_hour'	=> $guesses / (100.0 / 3600.0),
		'online_no_throttling_10_per_second'	=> $guesses / 10.0,
		'offline_slow_hashing_1e4_per_second'	=> $guesses / 1e4,
		'offline_fast_hashing_1e10_per_second'	=> $guesses / 1e10
	);
	
	my %crack_times_display = map { $_ => display_time($crack_times_seconds{$_}) } keys(%crack_times_seconds);
	
	return {
		'crack_times_seconds'	=> \%crack_times_seconds,
		'crack_times_display'	=> \%crack_times_display,
		'score'	=> guesses_to_score($guesses)
	};
}

1;


lib/ZXCVBN.pm
#!/usr/bin/perl

use 5.014;
use feature 'unicode_strings';
use strict;
use warnings;

use Carp qw();

package ZXCVBN;


=head1 NAME

ZXCVBN - A realistic password strength estimator

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

use ZXCVBN::Matching;
use ZXCVBN::Scoring;
use ZXCVBN::TimeEstimates;
use ZXCVBN::Feedback;

=head1 SYNOPSIS

ZXCVBN is a realistic password strength estimator

    use ZXCVBN;
    
    my $stats = zxcvbn('foo.password);

    my $z = ZXCVBN->new();
    
    my $stats2 = $z->check('another.foo.password');
    ...

=head1 EXPORT
=cut

use Exporter 'import';
our @EXPORT = qw(zxcvbn);

{
	my %RANKED_DICTIONARIES = ();
	
	sub zxcvbn($;\@) {
		my($password,$p_user_inputs) = @_;
		
		my $z = ZXCVBN->new(\%RANKED_DICTIONARIES);
		
		return $z->check($password,$p_user_inputs);
	}
}

=head1 SUBROUTINES/METHODS
=cut

sub new(;\%) {
	my $proto = shift;
	my $class = ref($proto) || $proto;
	
	my $p_RANKED_DICTIONARIES = shift;
	
	$p_RANKED_DICTIONARIES = {  }  unless(ref($p_RANKED_DICTIONARIES) eq 'HASH');
	
	my $self = {
		'RANKED_DICTIONARIES'	=>	$p_RANKED_DICTIONARIES
	};
	bless($self,$class);
	
	return $self;
}

sub check($;\@) {
	my $self = shift;
	
	Carp::croak((caller(0))[3].' is an instance method!')  if($^D > 0 && !ref($self));
	
	my($password,$p_user_inputs) = @_;
	
	$p_user_inputs = []  unless(defined($p_user_inputs));
	
	my $start = time();
	
	my @sanitized_inputs = ();
	foreach my $arg (@{$p_user_inputs}) {
		push(@sanitized_inputs,defined($arg) ? $arg.'' : undef);
	}
	
	my %ranked_dictionaries = %{$self->{'RANKED_DICTIONARIES'}};
	# User defined dictionary
	$ranked_dictionaries{'user_inputs'} = ZXCVBN::Matching::build_ranked_dict(@sanitized_inputs);
	
	my $p_matches = ZXCVBN::Matching::omnimatch($password,%ranked_dictionaries);
	
	my $p_result = ZXCVBN::Scoring::most_guessable_match_sequence($password, @{$p_matches});
	$p_result->{'calc_time'} = time() - $start;
	
	my $p_attack_times = ZXCVBN::TimeEstimates::estimate_attack_times($p_result->{'guesses'});
	
	#    for prop, val in attack_times.items():
	#	result[prop] = val
	
	@{$p_result}{keys(%{$p_attack_times})} = values(%{$p_attack_times});
	
	$p_result->{'feedback'} = ZXCVBN::Feedback::get_feedback($p_result->{'score'},@{$p_result->{'sequence'}});
	
	return $p_result;
}

=head2 function1

=cut

sub function1 {
}

=head2 function2

=cut

sub function2 {
}

=head1 AUTHOR

José María Fernández, C<< <jose.m.fernandez at bsc.es> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-zxcvbn-perl at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=ZXCVBN-Perl>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc ZXCVBN


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=ZXCVBN-Perl>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/ZXCVBN-Perl>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/ZXCVBN-Perl>

=item * Search CPAN

L<http://search.cpan.org/dist/ZXCVBN-Perl/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2017 José María Fernández.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this program; if not, write to the Free
Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA


=cut

1; # End of ZXCVBN
